crates/ra_assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -0,0 +1,331 @@
+    acc: &mut Assists,
+    ctx: &AssistContext,
+) -> Option<()> {
+    let variant = ctx.find_node_at_offset::<ast::EnumVariant>()?;
+    let field_list = match variant.kind() {
+        ast::StructKind::Tuple(field_list) => field_list,
+        _ => return None,
+    };
+    let variant_name = variant.name()?.to_string();
+    let variant_hir = ctx.sema.to_def(&variant)?;
+    if existing_struct_def(ctx.db, &variant_name, &variant_hir) {
+        return None;
+    }
+    let enum_ast = variant.parent_enum();
+    let enum_name = enum_ast.name()?.to_string();
+    let visibility = enum_ast.visibility();
+    let current_module_def =
+        ImportsLocator::new(ctx.db).find_imports(&enum_name).first()?.left()?;
+    let current_module = current_module_def.module(ctx.db)?;
+    let target = variant.syntax().text_range();
+    acc.add_in_multiple_files(
+        AssistId("extract_struct_from_enum_variant"),
+        "Extract struct from enum variant",
+        target,
+        |edit| {
+            let definition = Definition::ModuleDef(ModuleDef::EnumVariant(variant_hir));
+            let res = definition.find_usages(&ctx.db, None);
+            let start_offset = variant.parent_enum().syntax().text_range().start();
+            let mut visited_modules_set = FxHashSet::default();
+            visited_modules_set.insert(current_module);
+            for reference in res {
+                let source_file = ctx.sema.parse(reference.file_range.file_id);
+                update_reference(
+                    ctx,
+                    edit,
+                    reference,
+                    &source_file,
+                    &current_module_def,
+                    &mut visited_modules_set,
+                );
+            }
+            extract_struct_def(
+                edit,
+                enum_ast.syntax(),
+                &variant_name,
+                &field_list.to_string(),
+                start_offset,
+                ctx.frange.file_id,
+                &visibility,
+            );
+            let list_range = field_list.syntax().text_range();
+            update_variant(edit, &variant_name, ctx.frange.file_id, list_range);
+        },
+    )
+}
+
+fn existing_struct_def(db: &RootDatabase, variant_name: &str, variant: &EnumVariant) -> bool {
+    variant
+        .parent_enum(db)
+        .module(db)
+        .scope(db, None)
+        .into_iter()
+        .any(|(name, _)| name.to_string() == variant_name.to_string())
+}
+
+fn insert_import(
+    ctx: &AssistContext,
+    builder: &mut AssistBuilder,
+    path: &ast::PathExpr,
+    module: &Module,
+    module_def: &ModuleDef,
+    path_segment: ast::NameRef,
+) -> Option<()> {
+    let db = ctx.db;
+    let mod_path = module.find_use_path(db, module_def.clone());
+    if let Some(mut mod_path) = mod_path {
+        mod_path.segments.pop();
+        let use_path = format!("{}::{}", mod_path.to_string(), path_segment.to_string());
```

Comment:
We should stick to strongly-typed `ModPath`s here. To use `ModPath`, you need to get a hold of `Name`. You can get a name from `hir::Variant`: `let name = variant_hir.name(ctx.db);`